File: 数据结构/线性数据结构/可扩容的数组.c
#include <string.h>

struct IntArray {
	int size;			//数据长度
	int capacity;		//容器长度
	int * arr;			//首元素指针
};

int init(struct IntArray *, int);		/** 初始化 **/
int addFirst(struct IntArray *,int);	/** 添加元素到开头 **/
int addLast(struct IntArray *,int);		/** 添加元素到末尾 **/
int add(struct IntArray *, int, int);	/** 添加元素到指定的位置 **/
int set(struct IntArray *, int, int);	/** 设置元素到指定的位置 **/
void forEach(struct IntArray *);		/** 遍历,打印元素 **/
int indexOf(struct IntArray *,int);		/** 获取指定元素第一次出现的下标 **/
int removeIndex(struct IntArray *, int);/** 根据下标删除元素 **/
void removeItem(struct IntArray *, int);/** 删除第一个指定值的元素 **/
int removeFirst(struct IntArray *);		/** 删除第一个元素 **/
int removeLast(struct IntArray *);		/** 删除最后一个元素 **/


int init(struct IntArray *arr, int capacity) {
	if(capacity <= 0){
		return -1;
	}
	arr->capacity = capacity;
	arr->size = 0;
	arr->arr = (int *) calloc(sizeof(int), arr->capacity);
	return 0;
}

int addFirst(struct IntArray *arr,int value){
	return add(arr, 0, value);
}

int addLast(struct IntArray *arr,int value){
	return add(arr, arr->size, value);
}

int add(struct IntArray *arr, int index, int value) {
	if (index < 0 || index > arr->size) {
		return -1;
	}
	if (arr->size == arr->capacity) {
		arr->capacity = arr->capacity * 2;
		realloc(arr->arr, arr->capacity);
	}

	for (int i = arr->size - 1; i >= index; i--) {
		arr->arr[i + 1] = arr->arr[i];
	}
	arr->arr[index] = value;
	arr->size++;
	return 0;
}

int set(struct IntArray *arr, int index, int value) {
	if (index < 0 || index >= arr->size) {
		return 1;
	}
	arr->arr[index] = value;
	return 0;
}

int indexOf(struct IntArray *arr,int value){
	for (int i = 0; i < arr->size; i++) {
		if(arr->arr[i] == value){
			return i;
		}
	}
	return -1;
}

int removeIndex(struct IntArray *arr, int index) {
	if (index < 0 || index >= arr->size) {
		return -1;
	}
	int value = arr->arr[index];
	for (int i = index + 1; i < arr->size; i++) {
		arr->arr[i - 1] = arr->arr[i];
	}
	arr->size--;
	return value;
}

int removeFirst(struct IntArray *arr){
	return removeIndex(arr, 0);
}
int removeLast(struct IntArray *arr){
	return removeIndex(arr, arr->size - 1);
}

void removeItem(struct IntArray *arr, int value) {
	int index = indexOf(arr, value);
	if(index != -1){
		removeIndex(arr, index);
	}
}

void forEach(struct IntArray *arr) {
	printf("size=%d capacity=%d [", arr->size, arr->capacity);
	for (int i = 0; i < arr->size; i++) {
		printf("%d", arr->arr[i]);
		if (i != (arr->size - 1)) {
			printf(",");
		}
	}
	printf("]\n");
}

File: 数据结构/线性数据结构/可扩容的数组.java
package io.javaweb.datastructure;

import java.util.Arrays;
import java.util.function.Consumer;

public class Array<E> {

	private E[] arr = null;

	private int size;

	public Array() {
		this(10);
	}

	@SuppressWarnings("unchecked")
	public Array(int capacity) {
		this.arr = (E[]) new Object[capacity];
		this.size = 0;
	}

	public int getCapacity() {
		return this.arr.length;
	}

	public E get(int index) {
		if (index < 0 || index >= this.size) {
			throw new IllegalArgumentException("index 不能大于" + (this.size - 1) + ",不能小于0");
		}
		return this.arr[index];
	}

	public boolean contains(E item) {
		for (E e : this.arr) {
			if (e.equals(item)) {
				return true;
			}
		}
		return false;
	}

	public int indexOf(E item) {
		for (int i = 0; i < this.size; i++) {
			if (this.arr[i].equals(item)) {
				return i;
			}
		}
		return -1;
	}

	public void set(int index, E item) {
		if (index < 0 || index >= size) {
			throw new IllegalArgumentException("index 不能大于" + (this.size - 1) + ",不能小于0");
		}
		this.arr[index] = item;
	}

	public E remove(int index) {
		if (index < 0 || index >= this.size) {
			throw new IllegalArgumentException("index 不能大于" + (this.size - 1) + ",不能小于0");
		}
		E e = this.arr[index];
		for (int i = index + 1; i < size; i++) {
			this.arr[i - 1] = this.arr[i];
		}
		this.size--;
		return e;
	}

	public E removeFirst() {
		return this.remove(0);
	}

	public E removeLast() {
		return this.remove(this.size - 1);
	}

	public void removeElement(E item) {
		int index = this.indexOf(item);
		if (index != -1) {
			this.remove(index);
		}
	}

	public int size() {
		return this.size;
	}

	public void addFirst(E item) {
		this.add(0, item);
	}

	public void addLast(E item) {
		this.add(this.size, item);
	}

	//添加元素到指定的位置
	public void add(int index, E item) {
		if (index < 0 || index > this.size) {
			throw new IllegalArgumentException("index 不能大于" + (this.size - 1) + ",不能小于0");
		}
		if (this.size == this.arr.length) {
			//throw new IllegalArgumentException("数组满了");
			// 扩容
			this.resize(this.arr.length * 2);
		}
		for (int i = size - 1; i >= index; i--) {
			this.arr[i + 1] = this.arr[i];
		}

		this.arr[index] = item;
		this.size++;
	}

	public boolean isEmpty() {
		return this.size == 0;
	}

	public void consumer(Consumer<E> consumer) {
		if (this.size > 0) {
			for (E e : this.arr) {
				consumer.accept(e);
			}
		}
	}

	@Override
	public String toString() {
		StringBuilder stringBuilder = new StringBuilder();
		stringBuilder.append(String.format("Array: size = %d , capacity = %d\n", size, this.arr.length));
		stringBuilder.append('[');
		for (int i = 0; i < size; i++) {
			stringBuilder.append(this.arr[i]);
			if (i != size - 1) {
				stringBuilder.append(", ");
			}
		}
		stringBuilder.append(']');
		return stringBuilder.toString();
	}

	//执行扩容
	private void resize(int newCapacity) {
		this.arr = Arrays.copyOf(this.arr, newCapacity);
//		@SuppressWarnings("unchecked")
//		E[] newArr = (E[]) new Object[newCapacity];
//		for (int i = 0; i < size; i++) {
//			newArr[i] = this.arr[i];
//		}
//		this.arr = newArr;
	}
}
